server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -402,15 +408,18 @@ interface RespondingWriteResult {
      */
     static final class AsyncAfterWriteAction {
         private final Location location;
-        private final boolean waitUntilRefresh;
+        private final boolean needsRefreshAction;
         private final boolean sync;
         private final AtomicInteger pendingOps = new AtomicInteger(1);
         private final AtomicBoolean refreshed = new AtomicBoolean(false);
+        // TODO: Temporary until we fail unpromotable shard
```

Comment:
There can be more than one unpromotable shards, right?

If we want to fail the unpromotable shard(s), is there a reason we do not fail the shard in this PR? I am not sure yet how to recognize which shard failed. I wonder if the Exception has any field from which node it occurred. Else, we'd want to make `TransportBroadcastUnpromotableAction#doExecute()`'s response handler capture which node failed and pass it in the exception/response somehow.

Reply:
Yes. It will require multiple modifications and tests which is best left to a follow-up task.